### PR TITLE
feat: add tracing

### PR DESCRIPTION
**Pull request by lucasfernog-crabnebula**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [ ] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### CODE REVIEW

**Comment by CrabNejonas** [comment]
I like the name spacing! Just checked with the `tokio` repo and they use `::` (instead of `.`) for namespacing, should we do that as well for consistency?

**Comment by lucasfernog-crabnebula** [comment]
Of course you like the namespace, I copied it from your PR 😂 let's do what tokio is doing.

**Comment by CrabNejonas** [comment]
> Of course you like the namespace, I copied it from your PR 😂 let's do what tokio is doing.

oh lol, I just looked at my linear issue which doesn't have that but hey 😂 you're right it would explain that

File: core/tauri-macros/src/command/wrapper.rs [hunk 1]
```diff
@@ -211,7 +228,17 @@ pub fn wrapper(attributes: TokenStream, item: TokenStream) -> TokenStream {
           use #root::command::private::*;
           // prevent warnings when the body is a `compile_error!` or if the command has no arguments
           #[allow(unused_variables)]
-          let #root::ipc::Invoke { message: #message, resolver: #resolver } = $invoke;
+          let #root::ipc::Invoke { id: #id, message: #message, resolver: #resolver } = $invoke;
+
+          let _span = tracing::debug_span!(
+            "ipc::request::handler",
```
Comment (CrabNejonas) [inline core/tauri-macros/src/command/wrapper.rs]: Isn't this the top-level span for ipc requests? If so I would just use `ipc::request` for this and keep the `ipc::request::handler` for what's called `ipc::request::run` right now

```suggestion
            "ipc::request",
```

File: core/tauri-macros/src/command/wrapper.rs [hunk 2]
```diff
@@ -234,14 +261,22 @@ fn body_async(
...
+  let Invoke {
+    id,
+    message,
+    resolver,
+  } = invoke;
+  parse_args(id, function, message, attributes).map(|args| {
     quote! {
+      use tracing::Instrument;
+
+      let span = tracing::debug_span!("ipc::request::run", id = #id.0);
```
Comment (CrabNejonas) [inline core/tauri-macros/src/command/wrapper.rs]:
```suggestion
      let span = tracing::debug_span!("ipc::request::handler", id = #id.0);
```

See previous comment

File: core/tauri-macros/src/command/wrapper.rs [hunk 3]
```diff
@@ -267,6 +306,7 @@ fn body_blocking(
   });
 
   Ok(quote! {
+    let _span = tracing::debug_span!("ipc::request::run", id = #id.0).entered();
```
Comment (CrabNejonas) [inline core/tauri-macros/src/command/wrapper.rs]:
```suggestion
    let _span = tracing::debug_span!("ipc::request", id = #id.0).entered();
```

File: core/tauri-runtime-wry/src/lib.rs [hunk 4]
```diff
@@ -221,11 +221,35 @@ impl<T: UserEvent> Context<T> {
   }
 }
 
+#[derive(Debug, Clone, Default)]
+pub struct ActiveTraceSpanStore(Rc<RefCell<Vec<ActiveTracingSpan>>>);
```
Comment (CrabNejonas) [inline core/tauri-runtime-wry/src/lib.rs:244]: What is this used for?

File: core/tauri-runtime-wry/src/lib.rs [hunk 5]
```diff
@@ -2565,6 +2606,11 @@ fn create_webview<T: UserEvent, F: Fn(RawWindow) + Send + 'static>(
     ..
   } = pending;
 
+  let _webview_create_span = tracing::debug_span!("wry::webview::create").entered();
+  let window_draw_span = tracing::debug_span!("wry::window::draw").entered();
+  let window_create_span =
+    tracing::debug_span!(parent: &window_draw_span, "wry::window::create").entered();
```
Comment (CrabNejonas) [inline core/tauri-runtime-wry/src/lib.rs]: Do you think these might be better suited to live in the `wry` repo now that we have forked it too?  (in which case we wouldn't need the `wry::` prefix that we can derive from the span metadata)

File: core/tauri/src/app.rs [hunk 6]
```diff
@@ -356,26 +356,14 @@ impl<R: Runtime> AppHandle<R> {
...
-      )
-      .map_err(|e| crate::Error::PluginInitialization(plugin.name().to_string(), e.to_string()))?;
-    self
-      .manager()
-      .inner
-      .plugins
-      .lock()
-      .unwrap()
-      .register(plugin);
+  #[tracing::instrument(name = "app.plugin.register", skip(plugin), fields(name = plugin.name()))]
```
Comment (CrabNejonas) [inline core/tauri/src/app.rs]:
```suggestion
  #[tracing::instrument(name = "app::plugin::register", skip(plugin), fields(name = plugin.name()))]
```

File: core/tauri/src/app.rs [hunk 7]
```diff
@@ -808,6 +796,7 @@ shared_app_impl!(App<R>);
 shared_app_impl!(AppHandle<R>);
 
 impl<R: Runtime> App<R> {
+  #[tracing::instrument(name = "app.core_plugins.register")]
```
Comment (CrabNejonas) [inline core/tauri/src/app.rs]:
```suggestion
  #[tracing::instrument(name = "app::core_plugins::register")]
```

File: core/tauri/src/app.rs [hunk 8]
```diff
@@ -940,6 +929,7 @@ impl<R: Runtime> App<R> {
   /// }
   /// ```
   #[cfg(desktop)]
+  #[tracing::instrument(name = "app.run_iteration")]
```
Comment (CrabNejonas) [inline core/tauri/src/app.rs]:
```suggestion
  #[tracing::instrument(name = "app::run_iteration")]
```

File: core/tauri/src/app.rs [hunk 9]
```diff
@@ -1460,6 +1450,7 @@ impl<R: Runtime> Builder<R> {
 
   /// Builds the application.
   #[allow(clippy::type_complexity)]
+  #[tracing::instrument(name = "app.build", skip(self))]
```
Comment (CrabNejonas) [inline core/tauri/src/app.rs]:
```suggestion
  #[tracing::instrument(name = "app::build", skip(self))]
```

File: core/tauri/src/app.rs [hunk 10]
```diff
@@ -1686,6 +1677,7 @@ unsafe impl<R: Runtime> HasRawDisplayHandle for App<R> {
   }
 }
 
+#[tracing::instrument(name = "app.setup")]
```
Comment (CrabNejonas) [inline core/tauri/src/app.rs]:
```suggestion
#[tracing::instrument(name = "app::setup")]
```

File: core/tauri/src/app.rs [hunk 11]
```diff
@@ -1739,6 +1731,7 @@ fn setup<R: Runtime>(app: &mut App<R>) -> crate::Result<()> {
   Ok(())
 }
 
+#[tracing::instrument(name = "app.event_loop", skip(app_handle, callback, manager))]
```
Comment (CrabNejonas) [inline core/tauri/src/app.rs]:
```suggestion
#[tracing::instrument(name = "app::event_loop", skip(app_handle, callback, manager))]
```

File: core/tauri/src/ipc/protocol.rs [hunk 12]
```diff
@@ -275,6 +326,7 @@ fn handle_ipc_message<R: Runtime>(message: String, manager: &WindowManager<R>, l
         );
       }
       Err(e) => {
+        tracing::trace!("ipc.request.error {}", e);
```
Comment (CrabNejonas) [inline core/tauri/src/ipc/protocol.rs:351]:
```suggestion
        tracing::trace!("ipc::request::error {}", e);
```

File: core/tauri/src/ipc/protocol.rs [hunk 13]
```diff
@@ -42,11 +47,24 @@ pub fn get<R: Runtime>(manager: WindowManager<R>, label: String) -> UriSchemePro
...
+                tracing::trace_span!("ipc::request::handle", id = request.id.0, cmd = request.cmd);
+
+              let id = request.id;
+
               window.on_message(
                 request,
                 Box::new(move |_window, _cmd, response, _callback, _error| {
+                  let _span = tracing::trace_span!(
+                    parent: &request_span,
+                    "ipc::request::respond",
```
Comment (CrabNejonas) [inline core/tauri/src/ipc/protocol.rs]: This span ID is way to similar to `ipc::request::response`, It is also duplicating the data sent by the next span

File: core/tauri/src/lib.rs [hunk 14]
```diff
@@ -568,10 +568,32 @@ pub trait Manager<R: Runtime>: sealed::ManagerBase<R> {
   ///   app.emit_all("synchronized", ());
   /// }
   /// ```
+  #[tracing::instrument("app.emit.all", skip(self, payload))]
```
Comment (CrabNejonas) [inline core/tauri/src/lib.rs]:
```suggestion
  #[tracing::instrument("app::emit_all", skip(self, payload))]
```

File: core/tauri/src/lib.rs [hunk 15]
```diff
@@ -568,10 +568,32 @@ pub trait Manager<R: Runtime>: sealed::ManagerBase<R> {
...
+  /// ```
+  /// use tauri::Manager;
+  ///
+  /// #[tauri::command]
+  /// fn synchronize(app: tauri::AppHandle) {
+  ///   // emits the synchronized event to all windows
+  ///   app.emit_filter("synchronized", (), |w| w.label().starts_with("foo-"));
+  /// }
+  /// ```
+  #[tracing::instrument("app.emit.filter", skip(self, payload, filter))]
```
Comment (CrabNejonas) [inline core/tauri/src/lib.rs]:
```suggestion
  #[tracing::instrument("app::emit_filter", skip(self, payload, filter))]
```

File: core/tauri/src/lib.rs [hunk 16]
```diff
@@ -587,6 +609,7 @@ pub trait Manager<R: Runtime>: sealed::ManagerBase<R> {
   ///   }
   /// }
   /// ```
+  #[tracing::instrument("app.emit.to", skip(self, payload))]
```
Comment (CrabNejonas) [inline core/tauri/src/lib.rs]:
```suggestion
  #[tracing::instrument("app::emit_to", skip(self, payload))]
```

File: core/tauri/src/lib.rs [hunk 17]
```diff
@@ -651,6 +674,7 @@ pub trait Manager<R: Runtime>: sealed::ManagerBase<R> {
   ///   }
   /// }
   /// ```
+  #[tracing::instrument("app.emit.rust", skip(self))]
```
Comment (CrabNejonas) [inline core/tauri/src/lib.rs]:
```suggestion
  #[tracing::instrument("app::emit_rust", skip(self))]
```

I don't like that the span name differs from the method name so much

File: core/tauri/src/plugin.rs [hunk 18]
```diff
@@ -646,10 +655,25 @@ impl<R: Runtime> PluginStore<R> {
...
+        let _span = tracing::trace_span!("plugin::hooks::ipc", name = p.name()).entered();
         return p.extend_api(invoke);
       }
     }
     invoke.resolver.reject(format!("plugin {plugin} not found"));
     true
   }
 }
+
+#[tracing::instrument(name = "plugin.hooks.initialize", skip(plugin), fields(name = plugin.name()))]
```
Comment (CrabNejonas) [inline core/tauri/src/plugin.rs]:
```suggestion
#[tracing::instrument(name = "plugin::hooks::initialize", skip(plugin), fields(name = plugin.name()))]
```

File: core/tauri/src/window.rs [hunk 19]
```diff
@@ -371,6 +391,7 @@ impl<'a, R: Runtime> WindowBuilder<'a, R> {
   }
 
   /// Creates a new webview window.
+  #[tracing::instrument(name = "window.create")]
```
Comment (CrabNejonas) [inline core/tauri/src/window.rs]:
```suggestion
  #[tracing::instrument(name = "window::create")]
```

File: core/tauri/src/window.rs [hunk 20]
```diff
@@ -948,6 +971,7 @@ impl<R: Runtime> PartialEq for Window<R> {
 }
 
 impl<R: Runtime> Manager<R> for Window<R> {
+  #[tracing::instrument("window.emit.to", skip(self, payload))]
```
Comment (CrabNejonas) [inline core/tauri/src/window.rs]:
```suggestion
  #[tracing::instrument("window::emit_to", skip(self, payload))]
```

File: core/tauri/src/window.rs [hunk 21]
```diff
@@ -959,12 +983,14 @@ impl<R: Runtime> Manager<R> for Window<R> {
       .emit_filter(event, Some(self.label()), payload, |w| label == w.label())
   }
 
+  #[tracing::instrument("window.emit.all", skip(self, payload))]
```
Comment (CrabNejonas) [inline core/tauri/src/window.rs]:
```suggestion
  #[tracing::instrument("window::emit_all", skip(self, payload))]
```

File: core/tauri/src/window.rs [hunk 22]
```diff
@@ -2418,18 +2457,22 @@ impl<R: Runtime> Window<R> {
...
   }
 
-  pub(crate) fn emit_internal<S: Serialize>(
-    &self,
-    event: &str,
-    source_window_label: Option<&str>,
-    payload: S,
-  ) -> crate::Result<()> {
+  #[tracing::instrument(
+    "window.emit",
```
Comment (CrabNejonas) [inline core/tauri/src/window.rs]:
```suggestion
    "window::emit",
```

**Review by CrabNejonas — CHANGES_REQUESTED** [review]

**Comment by CrabNejonas** [comment]
Okay so general comments:

1. I think it makes sense to pull out the span IDs into constants if possible (deduplicate them from the binary and help with maintaining them later)
2. I can't check this without a proper span visualisation but I think in places you're not dropping spans guards before entering the next (doesn't matter right now though we can change that as we develop devtools)

File: core/tauri-macros/src/command/wrapper.rs (hunk 1 above)
Comment (lucasfernog-crabnebula) [reply core/tauri-macros/src/command/wrapper.rs]: it's not the top level, there's more logic in window itself before it gets here (plus right now the top-level span is in wry).

File: core/tauri-runtime-wry/src/lib.rs (hunk 4 above)
Comment (lucasfernog-crabnebula) [reply core/tauri-runtime-wry/src/lib.rs:244]: to track how much time it took to draw the window, so we keep the span alive for a while

File: core/tauri-runtime-wry/src/lib.rs (hunk 5 above)
Comment (lucasfernog-crabnebula) [reply core/tauri-runtime-wry/src/lib.rs]: I don't really want to bloat wry with a bunch of tracing calls, but whatever we all decide works for me. cc @Wu-Yu-Wei

**Comment by lucasfernog-crabnebula** [comment]
@CrabNejonas we do have this awesome span visualisation tool called devtools 😂

**Comment by lucasfernog-crabnebula** [comment]
I guess we should have a central module for span names right?

File: core/tauri-runtime-wry/src/lib.rs (hunk 5 above)
Comment (CrabNejonas) [reply core/tauri-runtime-wry/src/lib.rs]: I think adding tracing to wry makes sense too (and we'll get the request at some point anyway) as even users that use wry just as a library would want to get perf info

**Comment by CrabNejonas** [comment]
> I guess we should have a central module for span names right?

Yes or go straight with something like https://github.com/crabnebula-dev/devtools/pull/32

**Pull request closed without merging**